File: ui/src/migration/pveToken.ts
export const PVE_TOKEN_FORMAT = "user@realm!tokenid=secret";
export const PVE_TOKEN_EXAMPLE = "root@pam!nodal=xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx";

export function pveTokenError(token: string): string | null {
  const value = token.trim();
  if (!value) {
    return `Proxmox API token is required as ${PVE_TOKEN_FORMAT}. Example: ${PVE_TOKEN_EXAMPLE}`;
  }
  const at = value.indexOf("@");
  const bang = value.indexOf("!");
  const eq = value.lastIndexOf("=");
  if (at < 1 || bang < at + 2 || eq < bang + 2 || eq === value.length - 1) {
    if (at < 0 || bang < 0) {
      return `Proxmox API token must be ${PVE_TOKEN_FORMAT}, not the secret alone. Example: ${PVE_TOKEN_EXAMPLE}`;
    }
    return `Proxmox API token must be ${PVE_TOKEN_FORMAT}. Example: ${PVE_TOKEN_EXAMPLE}`;
  }
  const user = value.slice(0, at);
  const realm = value.slice(at + 1, bang);
  const id = value.slice(bang + 1, eq);
  const secret = value.slice(eq + 1);
  if (/\s/.test(user) || /\s|@|!/.test(realm) || /\s/.test(id) || /\s/.test(secret)) {
    return `Proxmox API token must be ${PVE_TOKEN_FORMAT}. Example: ${PVE_TOKEN_EXAMPLE}`;
  }
  return null;
}

File: internal/migration/pve_token.go
package migration

import (
	"fmt"
	"strings"
)

const (
	// PVETokenExample is the operator-visible form Proxmox shows once.
	PVETokenExample = "root@pam!nodal=xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
	// PVETokenFormat is user@realm!tokenid=secret.
	PVETokenFormat = "user@realm!tokenid=secret"
)

// ValidatePVEToken requires the full PVEAPIToken value. The UUID secret
// alone is not accepted because Authorization: PVEAPIToken= needs the
// user@realm!tokenid prefix.
func ValidatePVEToken(token string) error {
	token = strings.TrimSpace(token)
	if token == "" {
		return fmt.Errorf("proxmox API token is required as %s. Example: %s", PVETokenFormat, PVETokenExample)
	}
	at := strings.Index(token, "@")
	bang := strings.Index(token, "!")
	eq := strings.LastIndex(token, "=")
	if at < 1 || bang < at+2 || eq < bang+2 || eq == len(token)-1 {
		if at < 0 && bang < 0 {
			return fmt.Errorf("proxmox API token must be %s, not the secret alone. Example: %s", PVETokenFormat, PVETokenExample)
		}
		return fmt.Errorf("proxmox API token must be %s. Example: %s", PVETokenFormat, PVETokenExample)
	}
	user := token[:at]
	realm := token[at+1 : bang]
	id := token[bang+1 : eq]
	secret := token[eq+1:]
	if strings.ContainsAny(user, " \t") || strings.ContainsAny(realm, " \t@!") ||
		strings.ContainsAny(id, " \t") || strings.ContainsAny(secret, " \t") {
		return fmt.Errorf("proxmox API token must be %s. Example: %s", PVETokenFormat, PVETokenExample)
	}
	if realm == "" || id == "" || secret == "" {
		return fmt.Errorf("proxmox API token must be %s. Example: %s", PVETokenFormat, PVETokenExample)
	}
	return nil
}
